test/server/http_filter_base_test.cc

```diff
@@ -0,0 +1,89 @@
+#include "server/http_dynamic_delay_filter.h"
+#include "server/http_test_server_filter.h"
+#include "server/http_time_tracking_filter.h"
+
+#include "test/server/http_filter_integration_test_base.h"
+
+#include "gtest/gtest.h"
+
+namespace Nighthawk {
+
+using namespace testing;
+
+class HttpFilterBaseIntegrationTest
+    : public HttpFilterIntegrationTestBase,
+      public testing::TestWithParam<
+          std::tuple<Envoy::Network::Address::IpVersion, absl::string_view, bool>> {
+public:
+  HttpFilterBaseIntegrationTest() : HttpFilterIntegrationTestBase(std::get<0>(GetParam())){};
+};
+
+INSTANTIATE_TEST_SUITE_P(
+    IpVersions, HttpFilterBaseIntegrationTest,
+    ::testing::Combine(testing::ValuesIn(Envoy::TestEnvironment::getIpVersionsForTest()),
+                       testing::ValuesIn({absl::string_view(R"EOF(
+name: time-tracking
+typed_config:
+  "@type": type.googleapis.com/nighthawk.server.ResponseOptions
+  emit_previous_request_delta_in_response_header: "foo"
+)EOF"),
+                                          absl::string_view(R"EOF(
+name: dynamic-delay
+typed_config:
+  "@type": type.googleapis.com/nighthawk.server.ResponseOptions
+  static_delay: 0.1s
+)EOF"),
+                                          absl::string_view("name: test-server")}),
+                       testing::ValuesIn({true, false})));
+
+// Verify extensions are well-behaved when it comes to:
+// - GET requests.
+// - POST requests with an entity body (takes a slightly different code path).
+// - Valid but empty configuration.
+// - Bad request-level json configuration.
+// We will be low on functional expectations for the extensions, but this will catch hangs and
+// ensure that bad configuration handling is in-place.
+TEST_P(HttpFilterBaseIntegrationTest, BasicExtensionFlows) {
```

Comment:
Adding to what @wjuan-AFK commented on - it would really help if we can separate these into individual test cases - one per behavior. Such approach simplifies future code changes, because it ensures that the tests won't be affected when a new behavior is added or that less tests are affected when a behavior is changed. Additionally when a behavior breaks, we get a clear signal which behavior broke, rather than a generic failure saying "the BasicExtensionFlows no longer works".

I see your point about duplicating the parameter unpacking. I wonder if we could place that into a separate helper function. If we figure out we can't, I think we should prefer more instances of smaller independent tests for individual behavior to shorter code that uses parametrised tests. In other words, if we need to drop the parametrised test approach and write few more test cases, I think that is a worthwhile price to pay for shorter and easier to maintain test cases.

Can we burn another cycle on this and see if we can find a win-win solution?

Reply:
I will take a look. Hopefully a convenient helper will suffice to support this, and then I can split these out into different tests. I'd like to avoid the need to drop parameterized testing here if we can: we would not only have to duplicate the test(s) for our three current extensions, but also require that for future extensions that would want to piggy back on this.

Reply:
Done in 5ff6d569f61e314e4307aaf5285088d0ef014869.
@wjuan-AFK sorry, this was actually straight forward, and looking at the results, I rather like it better compared to what it was.